### DISCUSSION TITLE
k3s multiserver with etcd on LXC instances - Error adding a third master.

### DISCUSSION THREAD

**Discussion by maroci** [k3s-io/k3s#4232] [Q&A] [answered]
Hi, 
I cannot add a third master to a k3s clusters that has two masters two workers. 
I created LXC instances with images:ubuntu/bionic/amd64. 

The k3s cluster with two masters seems to work when I run a test with a simple nginx deployment.
Here the k3s cluster I created with two nodes:

- 2 instances lb with haproxy and keepalived
- 2 instances masters 
- 2 instances workers

However, when I stopped the master1 with a simple lxc stop master1, all cluster went down, for that reason I came back to my steps to add a third master. However, the k3s on the third master doesn't start.

**Command I used to create the master:**
curl -sfL https://get.k3s.io | INSTALL_K3S_EXEC="server --cluster-init" --disable servicelb --tls-san $lb_ip --node-taint node-role.kubernetes.io/master=true:NoSchedule sh -s -

Then I added the second node lb manually with sudo nano /etc/systemd/system/k3s.service

**Command I used to add the second master:**
curl -sfL https://get.k3s.io | INSTALL_K3S_EXEC="server --server https://10.254.34.149:6443 --disable servicelb --tls-san lb2 --tls-san $lb1 --node-taint node-role.kubernetes.io/master=true:NoSchedule" K3S_TOKEN=$token sh -s -

I configured haproxy and keepalived. Then I installed metallb.

Then I added two workers and create the nginx image to run a test:

> NAME       STATUS   ROLES                       AGE     VERSION
> master1s   Ready    control-plane,etcd,master   4h43m   v1.21.5+k3s2
> master2s   Ready    control-plane,etcd,master   3h57m   v1.21.5+k3s2
> worker1s   Ready    <none>                      3h44m   v1.21.5+k3s2
> worker2s   Ready    <none>                      3h43m   v1.21.5+k3s2

On my browser I can see the Nginx service running at the port 80 of the IP address assigned from metallb to the ngx svc. 

I run the same command I run for the master2, but k3s on the third master doesn't start.

On the third node:
**systemctl status k3s**

> ● k3s.service - Lightweight Kubernetes
>    Loaded: loaded (/etc/systemd/system/k3s.service; enabled; vendor preset: enabled)
>   Drop-In: /run/systemd/system/k3s.service.d
>            └─zzz-lxc-service.conf
>    Active: activating (auto-restart) (Result: exit-code) since Sat 2021-10-16 13:05:08 UTC; 3s ago
>      Docs: https://k3s.io
>   Process: 110560 ExecStart=/usr/local/bin/k3s server --server https://10.254.34.149:6443 --disable servicelb (code=exited, status=1/FAILURE)
>   Process: 110559 ExecStartPre=/sbin/modprobe overlay (code=exited, status=1/FAILURE)
>   Process: 110548 ExecStartPre=/sbin/modprobe br_netfilter (code=exited, status=1/FAILURE)
>   Process: 110544 ExecStartPre=/bin/sh -xc ! /usr/bin/systemctl is-enabled --quiet nm-cloud-setup.service (code=exited, status=0/SUCCESS)
>  Main PID: 110560 (code=exited, status=1/FAILURE)
> 
> Oct 16 13:05:08 master3S systemd[1]: k3s.service: Main process exited, code=exited, status=1/FAILURE
> Oct 16 13:05:08 master3S systemd[1]: k3s.service: Failed with result 'exit-code'.
> Oct 16 13:05:08 master3S systemd[1]: Failed to start Lightweight Kubernetes.

**journalctl -xe**

> Oct 16 13:15:24 master3S systemd[1]: k3s.service: Service hold-off time over, scheduling restart.
Oct 16 13:15:24 master3S systemd[1]: systemd-udev-trigger.service: Cannot add dependency job, ignoring: Unit systemd-udev-trigger.service is not loaded properly: Invalid argument.
Oct 16 13:15:24 master3S systemd[1]: k3s.service: Scheduled restart job, restart counter is at 186.
-- Subject: Automatic restarting of a unit has been scheduled
-- Defined-By: systemd
-- Support: http://www.ubuntu.com/support
-- 
-- Automatic restarting of the unit k3s.service has been scheduled, as the result for
-- the configured Restart= setting for the unit.
Oct 16 13:15:24 master3S systemd[1]: Stopped Lightweight Kubernetes.
-- Subject: Unit k3s.service has finished shutting down
-- Defined-By: systemd
-- Support: http://www.ubuntu.com/support
-- 
-- Unit k3s.service has finished shutting down.
Oct 16 13:15:24 master3S systemd[1]: Starting Lightweight Kubernetes...
-- Subject: Unit k3s.service has begun start-up
-- Defined-By: systemd
-- Support: http://www.ubuntu.com/support
-- 
-- Unit k3s.service has begun starting up.
Oct 16 13:15:24 master3S sh[119974]: + /usr/bin/systemctl is-enabled --quiet nm-cloud-setup.service
Oct 16 13:15:24 master3S sh[119974]: /bin/sh: 1: /usr/bin/systemctl: not found
Oct 16 13:15:24 master3S modprobe[119981]: modprobe: FATAL: Module br_netfilter not found in directory /lib/modules/5.4.0-88-generic
Oct 16 13:15:24 master3S modprobe[119989]: modprobe: FATAL: Module overlay not found in directory /lib/modules/5.4.0-88-generic
Oct 16 13:15:24 master3S k3s[119990]: time="2021-10-16T13:15:24.358759006Z" level=info msg="Starting k3s v1.21.5+k3s2 (724ef700)"
Oct 16 13:15:24 master3S k3s[119990]: time="2021-10-16T13:15:24.390588054Z" level=info msg="Managed etcd cluster not yet initialized"
Oct 16 13:15:24 master3S k3s[119990]: time="2021-10-16T13:15:24.404414787Z" level=info msg="Adding https://10.254.34.239:2380 to etcd cluster [master1S-e2536d91=https://10.254.34.149:23
Oct 16 13:15:24 master3S k3s[119990]: {"level":"warn","ts":"2021-10-16T13:15:24.405Z","caller":"clientv3/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"endp
Oct 16 13:15:24 master3S k3s[119990]: time="2021-10-16T13:15:24.405088946Z" level=fatal msg="starting kubernetes: preparing server: start managed database: joining etcd cluster: etcdser
Oct 16 13:15:24 master3S systemd[1]: k3s.service: Main process exited, code=exited, status=1/FAILURE
Oct 16 13:15:24 master3S systemd[1]: k3s.service: Failed with result 'exit-code'.
Oct 16 13:15:24 master3S systemd[1]: Failed to start Lightweight Kubernetes.
-- Subject: Unit k3s.service has failed
-- Defined-By: systemd
-- Support: http://www.ubuntu.com/support
-- 
-- Unit k3s.service has failed.
-- 
-- The result is RESULT.

I know it's a mess, I hope someone is going to give me a hint.

Thanks.

**Comment by brandond** [comment]
You've not included the full logs. Can you do `journalctl -u k3s &> k3s.log` and attach the resulting file?

**Comment by maroci** [comment]
Here the log file from Master 1 that seems to works

[k3s.log](https://github.com/k3s-io/k3s/files/7364086/k3s.log) 

And the log file from the Master 3 that doesn't want to join the cluster:

[k3smaster3S.log](https://github.com/k3s-io/k3s/files/7364023/k3smaster3S.log)

Thank you for answering me

**Comment by brandond** [comment, marked as answer]
It appears that you still have the joining node (master3s) joined to the cluster under a different IP address - 10.254.34.239, vs the previously registered IP address of 10.254.34.148:
```
Oct 18 10:21:29 master3S k3s[154142]: time="2021-10-18T10:21:29.203095821Z" level=info msg="Starting k3s v1.21.5+k3s2 (724ef700)"
Oct 18 10:21:29 master3S k3s[154142]: time="2021-10-18T10:21:29.234825490Z" level=info msg="Managed etcd cluster not yet initialized"
Oct 18 10:21:29 master3S k3s[154142]: time="2021-10-18T10:21:29.247206711Z" level=info msg="Adding https://10.254.34.239:2380 to etcd cluster [master1S-e2536d91=https://10.254.34.149:2380 master3S-b01675ad=https://10.254.34.148:2380 master2S-b963b4de=https://10.254.34.137:2380]"
Oct 18 10:21:29 master3S k3s[154142]: {"level":"warn","ts":"2021-10-18T10:21:29.247Z","caller":"clientv3/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"endpoint://client-8afa8f35-6c20-4277-b721-75cc6daf2622/10.254.34.149:2379","attempt":0,"error":"rpc error: code = Unavailable desc = etcdserver: unhealthy cluster"}
Oct 18 10:21:29 master3S k3s[154142]: time="2021-10-18T10:21:29.247942048Z" level=fatal msg="starting kubernetes: preparing server: start managed database: joining etcd cluster: etcdserver: unhealthy cluster"
```

Note that cluster member IP addresses should not change once they've been joined to the cluster.

Do you see `master3S` listed in `kubectl get nodes`? If so, use `kubectl delete node` to remove this node from the cluster so that it can be rejoined.

**Reply by maroci** [reply]
Thank you very very much brandond. 
I understand now and I fixed it.
